File: server/daw/capture_test.py
"""DAW capture, python half: the codecs and the engine's audio mixing. [DAWREC]

What is pinned here and nowhere else:

  * capture.py's FLAC round trip is lossless to the 24-bit floor (measured
    1.2e-7 worst error), and engine.py decodes the SAME file to the SAME
    floats — the encode scale and the decode scale are inverses across two
    files, which nothing but this test would notice drifting.
  * the region-seam proof EXTENDED to audio clips: a region render of a
    window is bit-identical to that window sliced out of a whole render,
    with a file-backed clip in the mix. Same argument as notes (per-sample
    addition, memoryless master), now proven, not argued.
  * the click puts its blips at exactly the samples it was told, and only
    there — the count-in the browser plays IS the timeline's arithmetic.

Runs under the rig venv (needs numpy; PyAV for the FLAC sections, which
skip honestly when it is missing — the wav fallback is then what's proven).
"""
import json
import os
import subprocess
import sys
import tempfile

import numpy as np

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)
import capture  # noqa: E402
import engine   # noqa: E402

passed = 0
failures = []


def ok(label, cond, detail=""):
    global passed
    if cond:
        passed += 1
        print(f"  ok    {label}")
    else:
        failures.append(label)
        print(f"  FAIL  {label}" + (f"\n          {detail}" if detail else ""))


SR = 48000
HAVE_AV = capture._have_av()
print(f"\n  (PyAV {'present' if HAVE_AV else 'ABSENT — flac sections degrade to the wav fallback'})")

print("\n  -- encode: raw f32 -> take file -> the exact floats back --")
with tempfile.TemporaryDirectory() as td:
    rng = np.random.default_rng(7)
    x = (rng.uniform(-0.9, 0.9, SR * 2)).astype(np.float32)
    raw = os.path.join(td, "cap.f32")
    x.tofile(raw)
    r = capture.encode({"sr": SR, "raw": raw, "out": os.path.join(td, "take.flac")})
    ok("encode answers ok with the sample count", r["ok"] and r["n_samples"] == SR * 2)
    ok(f"the format is honest ({r['format']})",
       r["format"] == ("flac" if HAVE_AV else "wav"))
    ok("the file exists under the name encode returned", os.path.getsize(r["out"]) > 0)

    # decode it back through capture.py (the comp flattener's reader)
    dec = capture.decode({"path": r["out"], "sr": SR, "out": os.path.join(td, "back.f32")})
    back = np.fromfile(dec["out"], dtype="<f4")
    ok("decode returns every sample", len(back) == len(x))
    err = float(np.max(np.abs(back.astype(np.float64) - x.astype(np.float64))))
    budget = 2.5e-7 if HAVE_AV else 0.0
    ok(f"round-trip error {err:.2e} is at the 24-bit floor (budget {budget:.1e})", err <= budget)

    # and through the ENGINE's reader — the render-side decode must agree
    y_engine = engine._read_audio_f64(r["out"], SR)
    ok("engine reads the same length", len(y_engine) == len(x))
    err2 = float(np.max(np.abs(y_engine - back.astype(np.float64))))
    ok(f"engine and capture decode IDENTICALLY (err {err2:.1e})", err2 == 0.0)

    rate_err = ""
    try:
        engine._read_audio_f64(r["out"], 44100)
    except ValueError as exc:
        rate_err = str(exc)
    ok("a rate mismatch is an ERROR naming the fix, never a resample",
       "re-import" in rate_err)

print("\n  -- the seam proof holds WITH a file-backed clip in the mix --")
with tempfile.TemporaryDirectory() as td:
    rng = np.random.default_rng(11)
    perf = (rng.uniform(-0.5, 0.5, SR)).astype(np.float32)   # a 1 s "performance"
    raw = os.path.join(td, "p.f32")
    perf.tofile(raw)
    r = capture.encode({"sr": SR, "raw": raw, "out": os.path.join(td, "tk.flac")})
    take_path = r["out"]

    notes = [
        {"inst": "pluck", "midi": 60, "vel": 100, "start_sample": 0,
         "dur_samples": SR // 2, "gain_db": 0, "seed": 1},
        {"inst": "drums", "midi": 36, "vel": 110, "start_sample": SR,
         "dur_samples": 100, "gain_db": 0, "seed": 2},
    ]
    audio = [{"path": take_path, "start_sample": int(SR * 0.75),
              "offset_samples": 0, "dur_samples": SR, "gain_db": -3.0}]
    whole = engine.render({"sr": SR, "start_sample": 0, "n_samples": SR * 2,
                           "notes": notes, "audio": audio,
                           "out": os.path.join(td, "whole.wav")})
    y_whole, _ = engine.read_wav_f32(os.path.join(td, "whole.wav"))
    # the same content as two abutting region windows
    y_regions = []
    for w0, n in ((0, SR), (SR, SR)):
        out = os.path.join(td, f"reg{w0}.wav")
        engine.render({"sr": SR, "start_sample": w0, "n_samples": n,
                       "notes": notes, "audio": audio, "out": out})
        y_regions.append(engine.read_wav_f32(out)[0])
    stitched = np.concatenate(y_regions)
    ok("region renders stitch BIT-IDENTICAL to the whole render (audio included)",
       len(stitched) == len(y_whole) and np.array_equal(stitched, y_whole))

    # placement: the clip's first audible sample is exactly start_sample
    silent = engine.render({"sr": SR, "start_sample": 0, "n_samples": SR * 2,
                            "notes": [], "audio": audio,
                            "out": os.path.join(td, "only.wav")})
    y_only, _ = engine.read_wav_f32(os.path.join(td, "only.wav"))
    nz = np.nonzero(np.abs(y_only) > 1e-6)[0]
    ok(f"the clip's first sample lands AT start_sample ({nz[0]} == {int(SR*0.75)})",
       len(nz) > 0 and nz[0] == int(SR * 0.75))
    ok("...and its last inside start+dur", nz[-1] < int(SR * 0.75) + SR)

    # offset/dur trim: a window into the middle of the file
    trimmed = [{"path": take_path, "start_sample": 100, "offset_samples": 1000,
                "dur_samples": 2000, "gain_db": 0.0}]
    engine.render({"sr": SR, "start_sample": 0, "n_samples": 4000,
                   "notes": [], "audio": trimmed, "out": os.path.join(td, "trim.wav")})
    y_trim, _ = engine.read_wav_f32(os.path.join(td, "trim.wav"))
    src = engine._read_audio_f64(take_path, SR)
    expect = np.tanh(0.7 * src[1000:3000]).astype(np.float32)
    ok("offset_samples/dur_samples trim exactly (bit-compare against the source)",
       np.array_equal(y_trim[100:2100], expect)
       and float(np.max(np.abs(y_trim[:100]))) == 0.0
       and float(np.max(np.abs(y_trim[2100:]))) == 0.0)


def _read_stereo(path):
    """The inverse of rack.write_wav_f32_stereo: its fixed 44-byte header, then
    interleaved float32. Local on purpose -- that writer IS the contract, and a
    chunk-walking reader would absorb a header change instead of failing on it."""
    with open(path, "rb") as fh:
        raw = fh.read()
    y = np.frombuffer(raw[44:], dtype="<f4")
    return np.vstack([y[0::2], y[1::2]])


print("\n  -- THE CHAINED PATH MIXES THE CLIP TOO (P0 is not the only lane) --")
#
# THE DEFECT THIS PINS SHUT. engine.render returned rack.render_with_chain the
# moment a job carried `mixer`, and rack.chain_graph built its dry buffers from
# job["notes"] and nothing else -- the word "audio" appeared nowhere in
# rack.py. store.js gives every NEW project master.stereo and stereo_on forces
# the chain path on it, so this was every new project with a recorded or
# imported take: the clip was hashed into the region identity, handed to the
# engine, and dropped. Measured here before the fix: the same one-second clip,
# peak 0.206966 through the P0 mono path and peak 0.0 through a NO-OP mixer,
# the chained render carrying the sha1 of the same job with no clip at all
# (ac92d2fd28154df1cb706af1c64c3908ce1e768b, both).
with tempfile.TemporaryDirectory() as td:
    import rack                                   # noqa: E402 -- the chain stage
    rng = np.random.default_rng(23)
    src = (rng.uniform(-0.3, 0.3, SR)).astype(np.float32)
    src_path = os.path.join(td, "clip.wav")
    engine.write_wav_f32(src_path, src, SR)
    CLIP = {"path": src_path, "start_sample": SR // 4, "offset_samples": 0,
            "dur_samples": SR, "gain_db": -3.0, "track_id": "t_a"}
    NOTES = [{"inst": "pluck", "midi": 60, "vel": 100, "start_sample": 0,
              "dur_samples": SR // 2, "gain_db": 0, "seed": 1, "track_id": "t_b"},
             {"inst": "drums", "midi": 36, "vel": 110, "start_sample": SR,
              "dur_samples": 100, "gain_db": 0, "seed": 2, "track_id": "t_b"}]
    MIX = {"stereo": True, "tracks": {"t_a": {}, "t_b": {}}, "returns": [],
           "master": {}, "spq": [[0.0, 0.5]]}
    BASE = {"sr": SR, "start_sample": 0, "n_samples": SR * 2}

    def _render(name, **kw):
        return engine.render(dict(BASE, out=os.path.join(td, name + ".wav"), **kw))

    chained = _render("chain_clip", notes=[], audio=[CLIP], mixer=MIX)
    bare = _render("chain_bare", notes=[], mixer=MIX)
    ok(f"a clip through a NO-OP mixer is AUDIBLE: peak {chained['peak']} (was 0.0)",
       chained["peak"] > 0.1 and chained["sha1"] != bare["sha1"])
    ok("...and the same job with no clip is still exactly silence, on the pinned sha1",
       bare["peak"] == 0.0
       and bare["sha1"] == "24fe969555a0783cba97f1f86fb1d604e9c7fc99")
    ok("both lanes count the clips they carried, in the same key",
       chained["clips"] == 1 and _render("p0_count", notes=[], audio=[CLIP])["clips"] == 1)

    # THE SEMANTICS ARE THE P0 PATH'S, TO THE BIT. A no-op chain is unity and
    # the pan law is centre-unity, so each channel of the chained render must
    # equal the mono render sample for sample. This is what "one implementation
    # of a clip into a buffer" means, measured rather than asserted:
    # rack._mix_audio calls engine.mix_clips, which is the code the line above
    # runs.
    p0 = _render("p0_clip", notes=[], audio=[CLIP])
    y_mono, _ = engine.read_wav_f32(os.path.join(td, "p0_clip.wav"))
    y_st = _read_stereo(os.path.join(td, "chain_clip.wav"))
    ok("chained L == the P0 mono render of the same clip, sample for sample",
       np.array_equal(y_st[0], y_mono))
    ok("...and L == R: a mono file lands in both channels, as a mono voice does",
       np.array_equal(y_st[0], y_st[1]))

    # THE SEAM PROOF, CARRIED THROUGH THE RACK. Same argument as the section
    # above: per-sample addition into a buffer that starts at absolute sample
    # 0, then a memoryless master, so a window IS a slice.
    parts = []
    for w0 in (0, SR):
        engine.render({"sr": SR, "start_sample": w0, "n_samples": SR, "notes": [],
                       "audio": [CLIP], "mixer": MIX,
                       "out": os.path.join(td, f"cw{w0}.wav")})
        parts.append(_read_stereo(os.path.join(td, f"cw{w0}.wav")))
    ok("region renders through the CHAIN stitch bit-identically to the whole "
       "render, with a clip in the mix",
       np.array_equal(np.hstack(parts), y_st))

    # BIT TRANSPARENCY, and it is the whole reason the placement was shared
    # rather than re-written. These sha1s were measured on the tree BEFORE the
    # chain lane could see a clip at all: every region cache on disk and the
    # region-seam proof are statements about exactly these bytes.
    ok("a job with NO clips renders byte-identically to before the fix, in every lane",
       _render("t_p0_notes", notes=NOTES)["sha1"]
       == "4e8f92db6aeece6aa42683e8f9555161a7d63761"
       and _render("t_ch_notes", notes=NOTES, mixer=MIX)["sha1"]
       == "7ab97c644ecd91e2b5813b47018109fb3c64eae1"
       and _render("t_ch_mono", notes=NOTES, mixer=dict(MIX, stereo=False))["sha1"]
       == "7ab97c644ecd91e2b5813b47018109fb3c64eae1")
    ok("...and the P0 lane's own clip renders did not move either (both pinned)",
       p0["sha1"] == "0d92a79a1e0dd6498be7f6d1d11071ffbf4127f5"
       and _render("t_p0_both", notes=NOTES, audio=[CLIP])["sha1"]
       == "6a45aa56a416a2c226800d0a633114a3ef839e3e")

    # A CLIP WITH NO track_id lands in the unnamed bus and is heard -- the same
    # rule _synth_notes reads a note's track by. Silence would be the one
    # answer this whole section exists to refuse.
    anon = dict(CLIP)
    anon.pop("track_id")
    ok("a clip with no track_id is still AUDIBLE (the unnamed bus, at unity)",
       _render("anon", notes=[], audio=[anon], mixer=MIX)["sha1"] == chained["sha1"])

    # THE STEMS. buses["tracks"] comes out of the same dry buffers, so a track
    # whose only content is a clip now has a lane -- and `silent_tracks` must
    # stop naming it, because voicelab.js's cache hit test is compared against
    # exactly that list on every render.
    st = rack.render_stems(dict(BASE, notes=NOTES, audio=[CLIP], mixer=MIX,
                                out_dir=os.path.join(td, "stems"), prefix="cp_"),
                           engine.SYNTHS, engine.TAILS)
    lanes = {s["track_id"]: s for s in st["stems"]}
    ok("render_stems writes a lane for the CLIP's track, not just the notes'",
       sorted(lanes) == ["t_a", "t_b"] and lanes["t_a"]["peak"] > 0.1,
       f"{sorted(lanes)}  silent={st['silent_tracks']}")
    ok("...and that track is no longer named silent", st["silent_tracks"] == [])
    ok("...and the lanes still sum to the pre-master mix, bit for bit",
       st["sums_to_mix"] is True and st["residual_db"] is None)

print("\n  -- the click puts blips exactly where it is told --")
with tempfile.TemporaryDirectory() as td:
    events = [{"sample": 0, "accent": True},
              {"sample": 12000, "accent": False},
              {"sample": 24000, "accent": False},
              {"sample": 47990, "accent": True}]      # deliberately near the end
    out = os.path.join(td, "click.wav")
    r = engine.click({"sr": SR, "n_samples": SR, "events": events, "out": out})
    ok("click reports every event placed", r["ok"] and r["clicks"] == 4)
    y, _ = engine.read_wav_f32(out)
    nz = np.nonzero(np.abs(y) > 1e-6)[0]
    # the blip is a sine, so its onset sample is a zero crossing: the first
    # NONZERO sample is event.sample + 1, and nothing sounds before it
    ok("the first audible sample is within 1 sample of event 0", len(nz) > 0 and nz[0] <= 1)
    gap = y[int(0.040 * SR) + 8:12000]                # after blip 0 decays, before blip 1
    ok("between blips is EXACT silence (no drift, no tail)",
       float(np.max(np.abs(gap))) == 0.0)
    ok("blip 2 starts exactly at its sample",
       float(np.max(np.abs(y[24000 - 40:24000]))) == 0.0 and abs(y[24000]) >= 0.0
       and np.any(np.abs(y[24000:24010]) > 1e-6))
    ok("a blip near the buffer end is clipped, not crashed", len(y) == SR)

if HAVE_AV:
    print("\n  -- decode: stereo 44.1k -> mono 48k (the import seam) --")
    with tempfile.TemporaryDirectory() as td:
        import av
        sr_in = 44100
        t = np.arange(sr_in) / sr_in
        left = (0.5 * np.sin(2 * np.pi * 440 * t)).astype(np.float32)
        right = (0.5 * np.sin(2 * np.pi * 660 * t)).astype(np.float32)
        src_path = os.path.join(td, "stereo.wav")
        cont = av.open(src_path, "w")
        st = cont.add_stream("pcm_f32le", rate=sr_in)
        st.codec_context.format = "flt"
        st.codec_context.layout = "stereo"
        inter = np.empty(sr_in * 2, dtype=np.float32)
        inter[0::2] = left
        inter[1::2] = right
        fr = av.AudioFrame.from_ndarray(inter.reshape(1, -1), format="flt", layout="stereo")
        fr.sample_rate = sr_in
        for pkt in st.encode(fr):
            cont.mux(pkt)
        for pkt in st.encode(None):
            cont.mux(pkt)
        cont.close()

        dec = capture.decode({"path": src_path, "sr": SR, "out": os.path.join(td, "m.f32")})
        ok("decode reports the source honestly", dec["src_sr"] == sr_in and dec["src_channels"] == 2)
        y = np.fromfile(dec["out"], dtype="<f4")
        ok(f"one second resamples to ~{SR} samples (got {len(y)})", abs(len(y) - SR) <= 64)
        ok("the downmix carries real signal", float(np.sqrt(np.mean(y ** 2))) > 0.1)

        p = capture.probe({"path": src_path})
        ok("probe names rate and channels", p["sr"] == sr_in and p["channels"] == 2)
else:
    print("\n  (skipping the stereo/resample section — PyAV absent; the wav "
          "fallback was proven above and decode of non-wav refuses honestly)")

print(f"\n  {passed} passed, {len(failures)} failed\n")
if failures:
    print("  failed:\n   " + "\n   ".join(failures) + "\n")
    sys.exit(1)

File: server/daw/capture_test.js
/**
 * DAW capture — the maths between "samples arrive" and "a take sits on the
 * timeline", proven closed-form. [DAWREC]
 *
 * Everything here has an exact answer: chunk reassembly is byte identity,
 * placement is one multiplication, a comp is picks over known buffers, a
 * count-in is bars × beats × seconds — so every assertion compares against
 * arithmetic, not against the code under test. The actor-honesty section
 * pins the one rule that must never drift: an agent driving the record API
 * cannot mint a human-performance provenance event.
 *
 * Runs standalone (`node server/daw/capture_test.js`) and in the pre-commit
 * hook. Touches no disk beyond one settings file in the OS temp dir.
 */
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { readFileSync } from "node:fs";
import { unlink } from "node:fs/promises";
import {
  beginSession, getSession, endSession, addChunk, assembleSession,
  latencyShift, takePlacement, punchTrim, flattenComp, clickEvents,
  readLatency, writeLatency, offsetFor, captureEvent, quantizePos,
} from "./capture.js";
import {
  blankProject, blankTrack, blankAudioClip, migrate,
  audioStartSample, audioEvents, regionHashes, regionsOf, dirtyBetween,
  posToSeconds, buildTimeline, SR, TICKS_PER_BEAT,
} from "./store.js";

let pass = 0;
const failures = [];

function ok(label, cond, detail = "") {
  if (cond) { pass++; console.log(`  ok    ${label}`); }
  else { failures.push(label); console.log(`  FAIL  ${label}${detail ? `\n          ${detail}` : ""}`); }
}
const near = (a, b, eps = 1e-9) => Math.abs(a - b) <= eps;
const thrown = (fn) => { try { fn(); return null; } catch (e) { return e.message; } };

function makeDoc(opts = {}) {
  const doc = blankProject("cap-test", { bpm: 120, num: 4, den: 4, lengthBars: 16, ...opts });
  doc.tracks.push(blankTrack("mic", "pluck"));
  return migrate(doc);
}

console.log("\n  -- chunk assembly is sample-exact across chunk boundaries --");
{
  const doc = makeDoc();
  const s = beginSession({
    slug: doc.slug, trackId: doc.tracks[0].id, sr: SR,
    at: { bar: 1, beat: 1, tick: 0 }, startSample: 0, shiftSamples: 0,
    countinBars: 1, countinSeconds: 2, device: "test", by: "user",
  });
  ok("the session is findable by id", getSession(s.id) === s);

  // a known ramp, split at deliberately awkward sizes (odd sample counts,
  // a 1-sample chunk) — reassembly must be the identical float sequence
  const total = 4801;
  const ramp = new Float32Array(total);
  for (let i = 0; i < total; i++) ramp[i] = Math.fround(Math.sin(i * 0.01) * (i / total));
  const cuts = [0, 1, 1024, 1025, 3000, 4800, total];
  const order = [3, 0, 5, 1, 4, 2];                 // arrival order ≠ seq order
  for (const seq of order) {
    const part = ramp.subarray(cuts[seq], cuts[seq + 1]);
    addChunk(s, seq, Buffer.from(part.buffer, part.byteOffset, part.byteLength));
  }
  const out = assembleSession(s);
  ok(`assembled length is exact (${out.length})`, out.length === total);
  let identical = true;
  for (let i = 0; i < total; i++) if (out[i] !== ramp[i]) { identical = false; break; }
  ok("every sample is bit-identical across the chunk seams", identical);

  ok("a duplicate seq is refused loudly", /already received/.test(thrown(() =>
    addChunk(s, 2, Buffer.alloc(8)))));
  ok("a ragged chunk (not ×4 bytes) is refused", /float32/.test(thrown(() =>
    addChunk(s, 6, Buffer.alloc(7)))));
  endSession(s.id);
  ok("a finished session is gone", /No such recording session/.test(thrown(() => getSession(s.id))));
}
{
  const s = beginSession({ slug: "x", trackId: "t", sr: SR, at: { bar: 1, beat: 1, tick: 0 },
    startSample: 0, shiftSamples: 0, countinBars: 0, countinSeconds: 0, by: "user" });
  addChunk(s, 0, Buffer.alloc(16));
  addChunk(s, 2, Buffer.alloc(16));
  ok("a MISSING seq fails assembly naming the gap", /chunk 1 is missing/.test(thrown(() => assembleSession(s))));
  endSession(s.id);
}

console.log("\n  -- the latency offset is applied at placement --");
{
  ok("50 ms at 48 kHz shifts 2400 samples EARLIER", latencyShift(50, SR) === -2400);
  ok("0 ms shifts nothing", latencyShift(0, SR) === 0);
  const doc = makeDoc();
  // bar 3 of 4/4 @120 starts at 4 s = 192000 samples
  const posSample = Math.round(posToSeconds(doc, { bar: 3, beat: 1, tick: 0 }) * SR);
  ok("bar 3 anchors at sample 192000", posSample === 192000);
  const s = beginSession({
    slug: doc.slug, trackId: doc.tracks[0].id, sr: SR,
    at: { bar: 3, beat: 1, tick: 0 }, startSample: posSample,
    shiftSamples: latencyShift(50, SR),
    countinBars: 1, countinSeconds: 2, by: "user",
  });
  const place = takePlacement(s);
  ok("the take lands 2400 samples before its anchor (189600)", place.startSample === 189600);
  endSession(s.id);

  // and store.audioStartSample re-derives the same number from the stored take
  const clip = blankAudioClip("tk_x.flac", { bar: 3, beat: 1, tick: 0, shiftSamples: -2400, durSamples: 1000 });
  ok("audioStartSample agrees with the placement formula",
    audioStartSample(doc, clip) === 189600);
}

console.log("\n  -- punch in/out is a pure trim --");
{
  const samples = new Float32Array(1000);
  for (let i = 0; i < 1000; i++) samples[i] = i;
  const t = punchTrim(samples, 5000, 5200, 5700);
  ok("the window is exact (500 samples from local 200)",
    t.samples.length === 500 && t.samples[0] === 200 && t.samples[499] === 699);
  ok("the start moved to the punch-in", t.startSample === 5200 && t.extraShift === 200);
  const noPunch = punchTrim(samples, 5000, null, null);
  ok("no punch = no trim", noPunch.samples.length === 1000 && noPunch.extraShift === 0);
  ok("an empty punch window is refused", /leaves nothing/.test(thrown(() =>
    punchTrim(samples, 5000, 7000, 8000))));
}

console.log("\n  -- comping: ordered picks flatten, later picks win --");
{
  const a = new Float32Array(400).fill(0.25);
  const b = new Float32Array(400).fill(0.5);
  const takes = new Map([
    ["tk_a", { start: 1000, samples: a }],
    ["tk_b", { start: 1200, samples: b }],
  ]);
  const flat = flattenComp([
    { take: "tk_a", fromSample: 1000, toSample: 1400 },
    { take: "tk_b", fromSample: 1300, toSample: 1600 },
  ], takes);
  ok("the comp window is min..max of the picks", flat.startSample === 1000 && flat.samples.length === 600);
  ok("before the overlap the first pick sounds", flat.samples[100] === 0.25);
  ok("in the overlap the LATER pick wins (order is priority)", flat.samples[350] === 0.5);
  ok("after take A ends only B sounds", flat.samples[550] === 0.5);
  // a pick past its take's samples leaves silence, honestly
  const flat2 = flattenComp([{ take: "tk_a", fromSample: 900, toSample: 1100 }], takes);
  ok("a pick reaching before the take leaves silence there",
    flat2.samples[0] === 0 && flat2.samples[99] === 0 && flat2.samples[100] === 0.25);
  ok("an unknown take is refused by name", /unknown take/.test(thrown(() =>
    flattenComp([{ take: "tk_zz", fromSample: 0, toSample: 10 }], takes))));
  ok("a backwards pick is refused", /to > from/.test(thrown(() =>
    flattenComp([{ take: "tk_a", fromSample: 10, toSample: 10 }], takes))));
}

console.log("\n  -- the count-in is meter-aware: 7/8 counts in 7 --");
{
  const doc = makeDoc();                               // 4/4 @120
  const ce = clickEvents(doc, 1, 4, 1, SR);
  ok("one 4/4 bar at 120 counts in over exactly 2 s", near(ce.countinSeconds, 2));
  ok("...which is 96000 samples", ce.countinSamples === 192000 / 2);
  const countinClicks = ce.events.filter((e) => e.sample < ce.countinSamples);
  ok("the count-in holds exactly 4 clicks", countinClicks.length === 4);
  ok("beat one of the count-in is accented, the rest are not",
    countinClicks[0].accent === true && countinClicks.slice(1).every((e) => !e.accent));
  ok("the first real bar's downbeat lands AT the count-in boundary",
    ce.events.some((e) => e.sample === ce.countinSamples && e.accent));

  const odd = makeDoc({ num: 7, den: 8 });             // 7/8 @120 quarter-bpm
  const ceOdd = clickEvents(odd, 1, 2, 1, SR);
  ok("one 7/8 bar at 120 counts in over 1.75 s", near(ceOdd.countinSeconds, 1.75));
  const oddCountin = ceOdd.events.filter((e) => e.sample < ceOdd.countinSamples);
  ok("SEVEN count-in clicks in 7/8 — never four", oddCountin.length === 7);
  ok("eighth-note spacing: clicks 0.25 s apart",
    oddCountin.every((e, i) => i === 0 || e.sample - oddCountin[i - 1].sample === SR / 4));

  // two count-in bars double the length
  const ce2 = clickEvents(doc, 1, 4, 2, SR);
  ok("countin_bars 2 counts 8 clicks over 4 s", near(ce2.countinSeconds, 4)
    && ce2.events.filter((e) => e.sample < ce2.countinSamples).length === 8);

  // a count-in into bar 5 uses BAR 5's meter
  const mixed = makeDoc();
  mixed.meterMap.push({ atBar: 5, num: 7, den: 8 });
  const ce5 = clickEvents(migrate(mixed), 5, 2, 1, SR);
  ok("counting into a 7/8 bar 5 counts in 7 (1.75 s), whatever bar 1 is",
    near(ce5.countinSeconds, 1.75)
    && ce5.events.filter((e) => e.sample < ce5.countinSamples).length === 7);
}

console.log("\n  -- audio clips reach the region hashes (the dirty mechanism) --");
{
  const doc = makeDoc();
  const before = regionHashes(doc);
  doc.tracks[0].audioClips.push(blankAudioClip("imp_1.flac", {
    bar: 9, beat: 1, tick: 0, durSamples: SR,        // one second in bar 9
  }));
  const after = regionHashes(doc);
  const dirty = dirtyBetween(before, after, regionsOf(doc));
  ok("adding a clip in bar 9 dirties exactly the bar-9 region",
    dirty.length === 1 && dirty[0].fromBar === 9 && dirty[0].toBar === 12, JSON.stringify(dirty));
  const evs = audioEvents(doc);
  ok("the event carries the absolute placement (bar 9 = 16 s = 768000)",
    evs.length === 1 && evs[0].startSample === 768000);
  const g1 = regionHashes(doc);
  doc.tracks[0].audioClips[0].gainDb = -6;
  const g2 = regionHashes(doc);
  ok("a clip gain change re-keys only its region",
    dirtyBetween(g1, g2, regionsOf(doc)).length === 1);
  doc.tracks[0].mute = true;
  ok("a muted track's audio clips vanish from the events (un-mute = dirty)",
    audioEvents(doc).length === 0);
}

console.log("\n  -- the per-device latency store --");
{
  const cfg = { settingsFile: path.join(os.tmpdir(), `dawrec_settings_${process.pid}.json`) };
  const t0 = await readLatency(cfg);
  ok("an absent settings file reads as an empty table", JSON.stringify(t0) === "{}");
  await writeLatency(cfg, "USB Mic (deluxe)", 87.312);
  await writeLatency(cfg, "default", 12);
  const t1 = await readLatency(cfg);
  ok("offsets round-trip per device label", t1["USB Mic (deluxe)"] === 87.312 && t1.default === 12);
  ok("offsetFor: exact label wins", offsetFor(t1, "USB Mic (deluxe)") === 87.312);
  ok("offsetFor: unknown label falls back to default", offsetFor(t1, "some other mic") === 12);
  ok("offsetFor: empty table is 0", offsetFor({}, "anything") === 0);
  await unlink(cfg.settingsFile).catch(() => {});
}

console.log("\n  -- ACTOR HONESTY at the capture seam --");
{
  const user = captureEvent("user", "audio", { device: "mic" });
  ok("a browser capture is a RECORD event (human-recorded)", user.type === "record");
  const agent = captureEvent("agent:mcp", "audio", {});
  ok("the SAME call driven by an agent is an IMPORT, never a record",
    agent.type === "import" && agent.data.origin === "third-party/existing");
  const sys = captureEvent("system", "audio", {});
  ok("an unattributable actor is an import too — never promoted", sys.type === "import");
  const midi = captureEvent("agent:mcp", "midi", {});
  ok("MIDI capture follows the same rule", midi.type === "import");
}

console.log("\n  -- quantize (optional, grid ticks) --");
{
  const q = quantizePos(2, 130, 240, 4);               // near beat 2 tick 240
  ok("130 ticks late snaps to the 240 grid", q.beat === 2 && q.tick === 240);
  const q2 = quantizePos(1, 100, 240, 4);
  ok("100 ticks snaps back to the beat", q2.beat === 1 && q2.tick === 0);
  const q3 = quantizePos(4, 900, 240, 4);              // would round past the bar
  ok("the bar's end clamps rather than spilling into bar+1",
    q3.beat === 4 && q3.tick <= TICKS_PER_BEAT - 1);
  ok("grid = a full beat snaps to beats", JSON.stringify(quantizePos(3, 500, 960, 4)) === JSON.stringify({ beat: 4, tick: 0 }));
}

/* The capture UI binds its listeners at MODULE TOP LEVEL, so one id that
 * daw.html does not carry is `null.addEventListener` — a TypeError during
 * evaluation that kills the WHOLE daw.js module, piano roll included, and
 * the browser swallows it. That is exactly what scripts/trace_load.mjs
 * exists for, and trace_load only evaluates web/app.js. A stub DOM cannot
 * catch it either (every stub element is a Proxy that answers to anything),
 * so the honest check is textual: every id daw.js reaches for must exist. */
console.log("\n  -- the capture UI is actually wired to elements that exist --");
{
  const web = path.join(path.dirname(fileURLToPath(import.meta.url)), "..", "..", "web");
  const js = readFileSync(path.join(web, "daw.js"), "utf8");
  const html = readFileSync(path.join(web, "daw.html"), "utf8");
  const inHtml = new Set([...html.matchAll(/id="([^"]+)"/g)].map((m) => m[1]));
  const wanted = [...new Set([...js.matchAll(/\$\("([^"]+)"\)/g)].map((m) => m[1]))];
  const missing = wanted.filter((id) => !inHtml.has(id));
  ok(`every id daw.js reaches for exists in daw.html (${wanted.length} checked)`,
    missing.length === 0, missing.join(", "));
  /* And the capture cluster by name, so deleting one of these from the page
   * fails here instead of removing a feature quietly. */
  for (const id of ["recBtn", "cntIn", "recDev", "impBtn", "impFile", "calOpen",
                    "midiDev", "midiRecBtn", "midiQuant", "takes",
                    "calDlg", "calRunMic", "calRunSyn", "calStore", "calResult"]) {
    ok(`the capture UI carries #${id}`, inHtml.has(id) && js.includes(`$("${id}")`));
  }
}

console.log(`\n  ${pass} passed, ${failures.length} failed\n`);
if (failures.length) {
  console.log("  failed:\n   " + failures.join("\n   ") + "\n");
  process.exit(1);
}
